hack/validate/vendor: print diff for modified files

kolyshkin opened the pull request:
In case some files were modified (rather than merely added
or removed), we're curious to see the diff for those.

Idea-by: @thaJeztah 

While at it,
* fix some shellcheck warnings;
* optimize/simplify `validate_vendor_used`.

kolyshkin committed:
hack/validate/vendor: print diff for modified files

In case some files were modified (rather than merely added
or removed), we're curious to see the diff for those.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
hack/validate/vendor: shellcheck fixes

The export statement is definitely not needed. The rest is obvious.

> In hack/validate/vendor line 3:
> export SCRIPTDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
>        ^-- SC2155: Declare and assign separately to avoid masking return values.
>
>
> In hack/validate/vendor line 43:
>	if ls -d vendor/$f  > /dev/null 2>&1; then
>                       ^-- SC2086: Double quote to prevent globbing and word splitting.
>
>
> In hack/validate/vendor line 44:
> 		found=$(find vendor/$f -iregex '.*LICENSE.*' -or -iregex '.*COPYRIGHT.*' -or -iregex '.*COPYING.*' | wc -l)
>                                   ^-- SC2086: Double quote to prevent globbing and word splitting.
>
>
> In hack/validate/vendor line 45:
>		if [ $found -eq 0 ]; then
>                    ^-- SC2086: Double quote to prevent globbing and word splitting.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

thaJeztah commented on hack/validate/vendor:
```diff
@@ -36,9 +40,9 @@ validate_vendor_diff(){
-		found=$(find vendor/$f -iregex '.*LICENSE.*' -or -iregex '.*COPYRIGHT.*' -or -iregex '.*COPYING.*' | wc -l)
-		if [ $found -eq 0 ]; then
+	if ls -d "vendor/$f"  > /dev/null 2>&1; then
+		found=$(find "vendor/$f" -iregex '.*LICENSE.*' -or -iregex '.*COPYRIGHT.*' -or -iregex '.*COPYING.*' | wc -l)
```
Not for this PR, wondering if those regexes could be combined (haven't tried), i.e.

```
'.*(LICENSE|COPYRIGHT|COPYING).*'
```

(guess "NOTICE" should also be in there)

kolyshkin replied:
I resisted the urge to rewrite this, yes. 

Alsop, `.*` at the beginning and the end is not needed.

kolyshkin replied:
added a commit optimizing it

thaJeztah reviewed (approved):
LGTM 👍

kolyshkin committed:
hack/validate/vendor: simplify looking for license

It was suggested that we use '.*\(COPYING\|LICENSE\|COPYRIGHT\).*'
as an argument to `find -iregex`, and this is how it all started.

Next thing, there is no COPYRIGHT in any of the vendored packages,
so it can be removed for good.

Next, we should not look too deep inside the package directory, as
the license should be in its root directory, so add `-maxdepth 1`
to `find`. This should also speed things up.

Finally, since we're not using the recursion feature of `find`,
it can be replaced with `echo | grep`.

While at it,
* avoid temporary $pkgs variable as it is only used once;
* replace `ls -d "vendor/$f"  > /dev/null 2>&1` with `test -d`.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
@thaJeztah first two commits remain the same (so you won't have to re-review), the only change is I added the third one with `validate_vendor_used` as you suggested.

kolyshkin commented:
> Should we add NOTICE? IIUC, the notice is the actual copyright/licensing information, and having a license file is merely for convenience

1. Currently, we don't have any packages that fail validation (i.e. they all either container COPYING or LICENSE in some way).

1. Let's have a look,

```console
$ grep -vEc '^$|^#' vendor.conf 
118
$ find vendor -name NOTICE\* | wc -l
13
```
...out of 118 packages that moby/moby currently vendors, only 13 packages contain NOTICE, and about half of those NOTICEs only contain a copyright statement but not any licensing information. I guess the aim of this check we are actually checking for licensing terms (to make sure it's compatible with moby's own licensing terms) not copyright.

thaJeztah reviewed (approved):
LGTM

andrewhsu reviewed (approved):
SGTM

would be nice to have an example run to show the effects of this change working; regardless, i already see the value of this PR in diagnosing certain situations.

SamWhited commented on hack/validate/vendor:
```diff
@@ -13,6 +13,7 @@ validate_vendor_diff(){
 		vndr
 		# check if any files have changed
 		diffs="$(git status --porcelain -- vendor 2>/dev/null)"
```
Should this check `vendor.conf` as well?

thaJeztah replied:
vendor.conf is only updated manually, so that shouldn't be an issue
